### PR TITLE
Enpassant

### PR DESCRIPTION
**Pull request by euruswagner**
1. My apologies for combining the two PR's. I kind of started and it all just went hand in hand. Luckily 90% is the same as the individual PR's from last week.
2. Biggest thing is I added two partials. One for each player. I then had the update action reload these partials also so that Castling and Captured pieces would show up without a full page reload. 
3. I added a few tests from the PR's I had last week to catch some situations I had missed. I missed that I was not checking that it had to be a pawn capturing a pawn.
4. Cut out some of the locals we had in the partials and update.js.erb since they were not needed.

### CODE REVIEW

File: app/controllers/pieces_controller.rb [hunk 1]
```diff
@@ -17,6 +17,13 @@ def update
     @piece.move_to!(x, y) if flash.now[:alert].empty?
   end
 
+  def castle
+    piece = Piece.find(params[:piece_id])
+    rook = Piece.find(params[:rook_id])
+    piece.castle!(rook)
+    redirect_to game_path(piece.game)
```
Comment (shanesfast) [inline app/controllers/pieces_controller.rb]: This would still trigger a full-page reload right?

File: app/controllers/pieces_controller.rb [hunk 2]
```diff
@@ -35,4 +42,16 @@ def valid_move?
...
+
+  def can_castle? 
+    piece = Piece.find(params[:piece_id])
+    rook = Piece.find(params[:rook_id])
+    if !piece.can_castle?(rook)
+      redirect_to game_path(piece.game), alert: "You can not Castle at this time!"
+    end
+
+    if piece.is_white? && piece.game.player_one != current_user || !piece.is_white? && piece.game.player_two != current_user
+      redirect_to game_path(piece.game), alert: "That is not your piece!"
```
Comment (shanesfast) [inline app/controllers/pieces_controller.rb]: I think this would full-page reload as well.

File: app/controllers/pieces_controller.rb [hunk 3]
```diff
@@ -35,4 +42,16 @@ def valid_move?
     y = params[:y_position].to_i
     @piece.valid_move?(x, y)
   end
+
+  def can_castle? 
+    piece = Piece.find(params[:piece_id])
+    rook = Piece.find(params[:rook_id])
+    if !piece.can_castle?(rook)
+      redirect_to game_path(piece.game), alert: "You can not Castle at this time!"
```
Comment (shanesfast) [inline app/controllers/pieces_controller.rb]: I think this would full-page reload as well.

File: app/views/pieces/update.js.erb [hunk 4]
```diff
@@ -1,2 +1,4 @@
-$('.game-board').html('<%= j (render partial: 'games/board', the_game: @game) %>');
+$('.game-board').html('<%= j (render partial: 'games/board') %>');
+$('.player-one').html('<%= j (render partial: 'games/player_one') %>');
```
Comment (shanesfast) [inline app/views/pieces/update.js.erb]: For clarity, you are wanting to do a full-page reload for castling but trigger the piece#update action for en_passant?

Comment (euruswagner) [reply app/views/pieces/update.js.erb]: I dont't know if I would call it a full page reload. But for the buttons for Castling to show up more then the board has to be reloaded after every move. The buttons could be moved to the board but then still have the problem of captured pieces not showing up. This seemed like the simplest (though not most optimal) solution.

Comment (shanesfast) [reply app/views/pieces/update.js.erb]: At this point it is a full-page reload for the castling method. It does not call update.js.erb. Which is fine, like you said we can change it later. I just wasn't sure if you had intended to only reload the player1/player2 columns for en passant only or not. The board and the columns are being reloaded after every move at this point already.

**Pull request merged**